BackendTaskPilot/Program.cs

```diff
@@ -0,0 +1,38 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+
+var builder = WebApplication.CreateBuilder(args);
+var app = builder.Build();
+
+ExchangeRateService service = new ExchangeRateService();
+
+// REST API to get exchange rates based on input currencies
+app.MapPost("/", service.GetExchangeRates);
+
+app.Run();
+
+public struct CnbRatesResponse
+{
+    [JsonPropertyName("rates")] public List<Rate> Rates { get; set; }
+}
+
+public struct Rate
+{
+    [JsonPropertyName("currencyCode")] public string CurrencyCode { get; set; }
+    [JsonPropertyName("rate")] public float RateValue { get; set; }
+}
+
+internal class ExchangeRateService
+{
+    private const string MainCurrenciesUrl = "https://api.cnb.cz/cnbapi/exrates/daily";
+
+    public IEnumerable<Rate> GetExchangeRates(List<string> currencies)
+    {
+        HttpClient client = new HttpClient();
+			
+        var data = client.GetStringAsync(MainCurrenciesUrl).Result;
+        var response = JsonSerializer.Deserialize<CnbRatesResponse>(data);
```

Comment:
A critical takeaway from here is to avoid using `.Result` and instead `await` the task
Learn why at [Avoiding simple mistakes in async await](https://www.anthonysteele.co.uk/AsyncBasicMistakes.html)

Check the in-memory cache first, and return rates if found, else get rates from provider

```csharp

```
```suggestion
// default the source currency if not provided
sourceCurrency ??= CurrencyCodes.CZK;

// Create the day to pass to external API. This makes it explicit that we are retrieving for today
var day = DateOnly.FromDateTime(DateTime.UtcNow);

// Handle errors and exceptions in lines below and return appropriate results

// get the exchange rate provider for source currency
var provider = await exchangeRateProviderFactory.Get(sourceCurrency);

// Get the exchange rates from the provider
var exchangeRates = provider.GetDailyExchangeRatesAsync(day, cancellationToken);
```